src/combinatorics/simple_graph.lean

```diff
@@ -0,0 +1,217 @@
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V)
+
+/-- `G.neighbor_set v` is the set of vertices adjacent to `v` in `G`. -/
+def neighbor_set (v : V) : set V := set_of (G.adj v)
+
+lemma ne_of_edge {a b : V} (hab : G.adj a b) : a ≠ b :=
+by { rintro rfl, exact G.loopless a hab }
+
+/--
+The edges of G consist of the unordered pairs of vertices related by
+`G.adj`.  It is given as a subtype of the symmetric square.
+-/
+def E : Type u := {x : sym2 V // x ∈ sym2.from_rel G.sym}
+
+/-- Allows us to refer to a vertex being a member of an edge. -/
+instance has_mem : has_mem V G.E := { mem := λ v e, v ∈ e.val }
+
+/-- Construct an edge from its endpoints. -/
+def edge_of_adj {v w : V} (h : G.adj v w) : G.E := ⟨⟦(v,w)⟧, h⟩
+
+lemma mem_of_adj {v w : V} (h : G.adj v w) :
+  v ∈ G.edge_of_adj h := sym2.mk_has_mem v w
+
+lemma mem_of_adj_right {v w : V} (h : G.adj v w) :
+  w ∈ G.edge_of_adj h := sym2.mk_has_mem_right v w
+
+lemma adj_iff_exists_edge {v w : V} (hne : v ≠ w) :
+  G.adj v w ↔ ∃ (e : G.E), v ∈ e ∧ w ∈ e :=
+begin
+  split, { intro, use ⟦(v,w)⟧, assumption, refine ⟨(G.mem_of_adj _), (G.mem_of_adj_right _)⟩ },
+  rintro ⟨e, ⟨w', hve⟩, ⟨v', hew⟩⟩,
+  have : e.val = ⟦(v,w)⟧, { rw [hve, sym2.eq_iff] at hew ⊢, cc },
+  have key := e.property, rwa this at key,
+end
+
+variables {G}
+
+/-- Given an edge and one vertex incident on it, construct the other one. -/
+noncomputable def E.other (e : G.E) {v : V} (h : v ∈ e) : V :=
+by { have : v ∈ e.val, apply h, exact this.other }
+
+lemma E.other_mem (e : G.E) {v : V} (h : v ∈ e) : e.other h ∈ e :=
+begin
+  change sym2.mem.other h ∈ e.val, have := sym2.mem_other_spec h,
+  convert sym2.mk_has_mem_right _ _; tauto,
+end
+
+lemma E.other_ne (e : G.E) {v : V} (h : v ∈ e) : e.other h ≠ v :=
+begin
+  have key := e.property,
+  erw [← sym2.mem_other_spec h, sym2.eq_swap] at key,
+  exact G.ne_of_edge key,
+end
+
+attribute [irreducible] E.other
+variables (G)
+
+instance E.inhabited [inhabited {p : V × V | G.adj p.1 p.2}] : inhabited G.E :=
+⟨begin
+  rcases inhabited.default {p : V × V | G.adj p.1 p.2} with ⟨⟨x, y⟩, h⟩,
+  use ⟦(x, y)⟧, rwa sym2.from_rel_prop,
+end⟩
+
+instance edges_fintype [decidable_eq V] [fintype V] [decidable_rel G.adj] :
+  fintype G.E := subtype.fintype _
+
+attribute [irreducible] E
+
+@[simp] lemma irrefl {v : V} : ¬G.adj v v := G.loopless v
+
+@[symm] lemma edge_symm (u v : V) : G.adj u v ↔ G.adj v u := ⟨λ x, G.sym x, λ x, G.sym x⟩
+
+@[simp] lemma mem_neighbor_set (v w : V) : w ∈ G.neighbor_set v ↔ G.adj v w :=
+by tauto
+
+section finite_at
+
+/-!
+## Finiteness at a vertex
+
+This section contains definitions and lemmas concerning vertices that
+have finitely many adjacent vertices.  We denote this condition by
+`fintype (G.neighbor_set v)`.
+
+We define `G.neighbor_finset v` to be the `finset` version of `G.neighbor_set v`.
+Use `neighbor_finset_eq_filter` to rewrite this definition as a `filter`.
+-/
+
+variables (v : V) [fintype (G.neighbor_set v)]
```

Comment:
Why not avoid the coercion and use `set.finite` here? It seems a bit unpleasant to define this as a `set` but then use `fintype`.

Reply:
This lets `G.degree v` work without needing to pass a finiteness proof.  For example, the instance `neighbor_set_fintype` a bit later on means `G.degree v` can find all the `fintype` instances it needs for finite graphs.

Also, isn't `set.fintype s` defined as `nonempty (fintype s)`?  One goal for the graph library is to be able to eventually prove basic graph algorithms are correct, and it seems like using `nonempty` might make things noncomputable.